### PR TITLE
feat: processAction: 'sendWith' and 'noSend' 

### PR DESCRIPTION
**Pull request by chris-4chain**
## Description of Changes

Provide a brief description of the changes you've made.

## Linked Issues / Tickets

Reference any related issues or tickets, e.g. "Closes #123".

## Testing Procedure

Describe the tests you've added or any testing steps you've taken.

- [ ] I have added new unit tests
- [ ] All tests pass locally
- [ ] I have tested manually in my local environment

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have run the linter

### CODE REVIEW

File: pkg/storage/internal/actions/process.go [hunk 1]
```diff
@@ -45,22 +45,39 @@ func newProcessAction(logger *slog.Logger, txRepo TransactionsRepo, commissionCf
...
 	}
 
-	if args.IsDelayed {
-		panic("not implemented yet")
+	return p.broadcastTxs(ctx, p.txIDsToBroadcast(args))
+}
+
+func (p *process) txIDsToBroadcast(args *wdk.ProcessActionArgs) []string {
+	count := len(args.SendWith)
+	if args.TxID != nil {
```
Comment (dorzepowski) [inline pkg/storage/internal/actions/process.go:68]: AFAIR this should be included only if NoSend is false

File: pkg/storage/internal/actions/process.go [hunk 2]
```diff
@@ -45,22 +45,39 @@ func newProcessAction(logger *slog.Logger, txRepo TransactionsRepo, commissionCf
...
+		count++
 	}
 
-	return p.broadcastSingleTx(ctx, string(*args.TxID))
+	result := make([]string, 0, count)
+	for _, txID := range args.SendWith {
+		result = append(result, string(txID))
+	}
+	if args.TxID != nil {
```
Comment (dorzepowski) [inline pkg/storage/internal/actions/process.go:76]: AFAIR this should be included only if NoSend is false

File: pkg/storage/internal/actions/process.go [hunk 3]
```diff
@@ -215,24 +232,33 @@ func (p *process) newStatuses(args *wdk.ProcessActionArgs) (txStatus wdk.TxStatu
 	return
 }
 
-func (p *process) broadcastSingleTx(ctx context.Context, txID string) (*wdk.ProcessActionResult, error) {
-	sendStatus, err := p.getSendStatus(ctx, txID)
+func (p *process) broadcastTxs(ctx context.Context, txIDs []string) (*wdk.ProcessActionResult, error) {
+	if len(txIDs) == 0 {
```
Comment (dorzepowski) [inline pkg/storage/internal/actions/process.go]: this shouldn't ever happen - we can skip this check here

File: pkg/internal/storage/repo/known_tx.go [hunk 4]
```diff
@@ -135,17 +143,25 @@ func (p *KnownTx) FindKnownTxRawTxs(ctx context.Context, txIDs []string) (map[st
...
-	return model.Status, nil
+
+	return seq.Reduce(
+		seq.FromSlice(rows),
+		func(acc map[string]wdk.ProvenTxReqStatus, row *models.KnownTx) map[string]wdk.ProvenTxReqStatus {
+			acc[row.TxID] = row.Status
+			return acc
+		},
+		make(map[string]wdk.ProvenTxReqStatus, len(rows)),
+	), nil
```
Comment (dorzepowski) [inline pkg/internal/storage/repo/known_tx.go]:
```suggestion
	txIDStatuses := seq.MapTo(seq.FromSlice(rows), func(row *models.KnownTx) (string, wdk.ProvenTxReqStatus) {
		return row.TxID, row.Status
	})
	
	return seq2.CollectToMap(txIDStatuses), nil
```

File: pkg/internal/storage/repo/known_tx.go [hunk 5]
```diff
@@ -261,7 +277,7 @@ func (p *KnownTx) recursiveBuildValidBEEF(ctx context.Context, depth int, mergeT
 	return nil
 }
 
-func (p *KnownTx) GetBEEFForTxIDs(ctx context.Context, txids iter.Seq[string], knownTxIDs []string, statusesToFilterOut []wdk.ProvenTxReqStatus) ([]byte, error) {
+func (p *KnownTx) BuildValidBEEFForTxIDs(ctx context.Context, txids iter.Seq[string], knownTxIDs []string, statusesToFilterOut []wdk.ProvenTxReqStatus) (*transaction.Beef, error) {
```
Comment (dorzepowski) [inline pkg/internal/storage/repo/known_tx.go]: ### Suggestion

What does it mean `Valid`?
Why it need a name `Build`.

In other words, IMHO previous name was better when looking from domain perspective.
Maybe in other repository we won't need to build it, we would just select it with query (imagine we have neo4j attached)

File: pkg/storage/internal/actions/process.go [hunk 6]
```diff
@@ -215,24 +232,33 @@ func (p *process) newStatuses(args *wdk.ProcessActionArgs) (txStatus wdk.TxStatu
...
+	}
+
+	sendStatusesLookup, err := p.getSendStatuses(ctx, txIDs...)
 	if err != nil {
 		return nil, err
 	}
 
-	if sendStatus != wdk.SendWithResultStatusSending {
+	sendStatuses := maps.Values(sendStatusesLookup)
+	if seq.ContainsAll(sendStatuses, wdk.SendWithResultStatusUnproven) {
```
Comment (dorzepowski) [inline pkg/storage/internal/actions/process.go]: Are you sure this is what you want to use?
Shouldn't It be rather:
```suggestion
	if seq.Every(sendStatuses, is.EqualTo(wdk.SendWithResultStatusUnproven)) {
```

Otherwise we won't broadcast any transaction if at least one would have status "unproven"

File: pkg/storage/internal/actions/process.go [hunk 7]
```diff
@@ -215,24 +232,33 @@ func (p *process) newStatuses(args *wdk.ProcessActionArgs) (txStatus wdk.TxStatu
...
-			SendWithResults: []wdk.SendWithResult{
-				{
+			SendWithResults: seq.Collect(seq2.MapTo(maps.All(sendStatusesLookup), func(txID string, status wdk.SendWithResultStatus) wdk.SendWithResult {
+				return wdk.SendWithResult{
 					TxID:   primitives.TXIDHexString(txID),
-					Status: sendStatus,
-				},
-			},
+					Status: status,
+				}
```
Comment (dorzepowski) [inline pkg/storage/internal/actions/process.go]: shouldn't we also add "NotDelayedResults"?

File: pkg/storage/internal/actions/process.go [hunk 8]
```diff
@@ -243,30 +269,54 @@ func (p *process) broadcastSingleTx(ctx context.Context, txID string) (*wdk.Proc
...
-	if !ok {
-		return nil, fmt.Errorf("failed to find aggregated result for txID %s", txID)
-	}
+	sendWithResults := make([]wdk.SendWithResult, 0, len(txIDs))
+	notDelayedResults := make([]wdk.ReviewActionResult, 0, len(txIDs))
+
+	aggregated := results.Aggregated(txIDs)
+	for _, txID := range txIDs {
+		aggBroadcastResult, ok := aggregated[txID]
+		if !ok {
```
Comment (dorzepowski) [inline pkg/storage/internal/actions/process.go:296]: (this comment is valid only if the comment about seq.ContainsAll(sendStatuses, wdk.SendWithResultStatusUnproven) is valid )


maybe I'm wrong, but if one txID would be "unproven" then we would return here error, for that transaction

File: pkg/storage/internal/actions/process.go [hunk 9]
```diff
@@ -45,22 +45,39 @@ func newProcessAction(logger *slog.Logger, txRepo TransactionsRepo, commissionCf
 }
 
 func (p *process) Process(ctx context.Context, userID int, args *wdk.ProcessActionArgs) (*wdk.ProcessActionResult, error) {
```
Comment (dorzepowski) [inline pkg/storage/internal/actions/process.go:47]: It would be good to add some tests (especially for corner cases) for ProcessAction on storage with NoSend and SendWith

File: pkg/storage/internal/actions/process.go [hunk 10]
```diff
@@ -215,24 +232,33 @@ func (p *process) newStatuses(args *wdk.ProcessActionArgs) (txStatus wdk.TxStatu
...
 		return nil, err
 	}
 
-	if sendStatus != wdk.SendWithResultStatusSending {
+	sendStatuses := maps.Values(sendStatusesLookup)
+	if seq.ContainsAll(sendStatuses, wdk.SendWithResultStatusUnproven) {
 		return &wdk.ProcessActionResult{
-			SendWithResults: []wdk.SendWithResult{
-				{
+			SendWithResults: seq.Collect(seq2.MapTo(maps.All(sendStatusesLookup), func(txID string, status wdk.SendWithResultStatus) wdk.SendWithResult {
```
Comment (dorzepowski) [inline pkg/storage/internal/actions/process.go]: ### minor thing

there are multiple places where `maps.All(sendStatusesLookup)` is called
you could create a variable just after p.getSendStatuses(ctx, txIDs...) as it will be used in all places.

Or you could use `seq2.FromMap` as this is the same - and would keep the code in the same namespace

File: pkg/storage/internal/actions/process.go [hunk 11]
```diff
@@ -243,30 +269,54 @@ func (p *process) broadcastSingleTx(ctx context.Context, txID string) (*wdk.Proc
...
+		}
+
+		err = p.knownTxRepo.UpdateKnownTxStatus(ctx, txID, newReqStatus, wdk.ProvenTxReqBeyondBroadcastStageStatuses, notes)
+		if err != nil {
+			return nil, fmt.Errorf("failed to update transaction status after broadcast: %w", err)
+		}
+
+		sendWithResults = append(sendWithResults, sendWithResult)
+		notDelayedResults = append(notDelayedResults, reviewActionResult)
 	}
```
Comment (dorzepowski) [inline pkg/storage/internal/actions/process.go]: It would be nice to have method for creation of result from post beef result

File: pkg/storage/internal/actions/process.go (hunk 1 above)
Comment (chris-4chain) [reply pkg/storage/internal/actions/process.go:68]: According to the TS version: `no`.
If `args.IsNoSend && len(args.SendWith) == 0`
then we don't reach this point

Otherwise, if `SendWith` is not empty, the `IsNoSend` is ignored and all the `SendWith` txid and the "new transaction" as well are sent.
See the "note" comment, I made:
> // NOTE: SendWith overrides IsNoSend, so if SendWith is NOT empty, we will broadcast txs anyway

File: pkg/storage/internal/actions/process.go (hunk 2 above)
Comment (chris-4chain) [reply pkg/storage/internal/actions/process.go:76]: ☝️

File: pkg/storage/internal/actions/process.go (hunk 3 above)
Comment (chris-4chain) [reply pkg/storage/internal/actions/process.go]: I'm not so sure. And if the implementator is not sure - I think, some "defensive programming" is acceptable. 
Besides, now maybe it won't happen, but it can be small change in validation when this starts to be necessary.

File: pkg/internal/storage/repo/known_tx.go (hunk 5 above)
Comment (chris-4chain) [reply pkg/internal/storage/repo/known_tx.go]: This is legacy from TS naming. But you have valid point.
I'll rename `BuildValidBEEFForTxIDs` -> `GetBEEFForTxIDs` and `BuildValidBEEFForTxID` -> `GetBEEFForTxID`

File: pkg/storage/internal/actions/process.go (hunk 6 above)
Comment (chris-4chain) [reply pkg/storage/internal/actions/process.go]: Good point. I misunderstood the `ContainsAll` function purpose.

File: pkg/storage/internal/actions/process.go (hunk 7 above)
Comment (chris-4chain) [reply pkg/storage/internal/actions/process.go]: According to `TS` version: `no`.
In there, the `ndr` (Not Delayed Result) field is added **after** the check:
```js
if (readyToSendReqIds.length < 1) {
    return createResults()
  }
```

File: pkg/storage/internal/actions/process.go (hunk 8 above)
Comment (chris-4chain) [reply pkg/storage/internal/actions/process.go:296]: I've checked `ContainsAll` -> `Every`
- So I guess, this comment in not applicable.

File: pkg/storage/internal/actions/process.go (hunk 9 above)
Comment (chris-4chain) [reply pkg/storage/internal/actions/process.go:47]: It's not so easy - that's why I decided to test it from "wallet" side. 
This way, it covers most of cases. 

But sure, we could discuss it, because it would require more tweaks.

File: pkg/storage/internal/actions/process.go (hunk 10 above)
Comment (chris-4chain) [reply pkg/storage/internal/actions/process.go]: I did it, but for me the variable name was less readable than explicit `maps.All(sendStatusesLookup)`.
And since this doesn't have any influence on the performance, I gave up.

File: pkg/storage/internal/actions/process.go (hunk 11 above)
Comment (chris-4chain) [reply pkg/storage/internal/actions/process.go]: I refactored this part a little bit - check it out 👍

**Review by mgosek-4chain — APPROVED** [review]

File: pkg/storage/internal/actions/process.go (hunk 1 above)
Comment (dorzepowski) [reply pkg/storage/internal/actions/process.go:68]: my mistake

File: pkg/storage/internal/actions/process.go (hunk 3 above)
Comment (dorzepowski) [reply pkg/storage/internal/actions/process.go]: it's a job for Superman .... no, Tests

File: pkg/storage/internal/actions/process.go (hunk 8 above)
Comment (dorzepowski) [reply pkg/storage/internal/actions/process.go:296]: actually it is if ContainsAll -> Every

File: pkg/storage/internal/actions/process.go (hunk 9 above)
Comment (dorzepowski) [reply pkg/storage/internal/actions/process.go:47]: As we talked, it will be in another PR

File: pkg/storage/internal/actions/process.go (hunk 3 above)
Comment (chris-4chain) [reply pkg/storage/internal/actions/process.go]: ok, ok 😅 - removed. More tests will be added when we create possibility to "sign" transactions on storage level for tests

File: pkg/storage/internal/actions/process.go (hunk 7 above)
Comment (dorzepowski) [reply pkg/storage/internal/actions/process.go]: maybe to be consistent, let's return NotDelayedResults always

**Review by dorzepowski — APPROVED** [review]
LGETM
